### PR TITLE
Handle download count round up to 1E+3K for when 999,500 <= count <=999,999

### LINKED ISSUE #8800: PM UI does not humanize the download count when 999,500 <= count <=999,999
image

### LINKED ISSUE #10864: NuGet packages downloaded over 1 billion times show the wrong unit
image

_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/NuGet-packages-downloaded-over-1-billion/1418778)._

---
[severity:It bothers me. A fix would be nice]
I was just about to download Newtonsoft.Json and noticed that the NuGet package apparently has 1.02G downloads, isn't this supposed to be 1.02B?

---
#### Feedback Bot on 5/11/2021, 11:14 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by erdembayar**
## Bug

Fixes: https://github.com/NuGet/Home/issues/8800
Fixes: https://github.com/NuGet/Home/issues/10864

Regression? Last working version:

## Description
For number higher than 999.5 fractional number(it's less than 1000) it round up to 1000 but it ends up 1E+3K, 1E+3M, 1E+3G text. This problem can be solved in many different ways, but I selected current approach which introduce least breaking change on UI screen, my change affect 0.1%  ~ 1/1000 cases, so 999/1000 cases users wouldn't notice it any difference. 

---
### Update 
Added localization after PR review comment. Now english users see `1B` instead of `1G`, but depending on localization other users see differently.

Previous experience on PMUI:
image

image

New experience on PMUI:
image

image

`Original number` - `Previous experience`  - `New experience`
image

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs [hunk 1]
```diff
@@ -64,15 +64,17 @@ public static string NumberToString(long number, IFormatProvider culture)
 
             while (v >= 1000)
             {
-                v /= 1000;
+                v = Math.Round(v / 1000, 2);
                 ++exp;
             }
 
+            string format = v < 999 ? "{0:G3}{1}" : "{0:G4}{1}"; // If number is 999561 then v = 999.56 so with G3 which result in 1000k ~ 1E+03K, but with G4 it result in 999.6K.
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs]: I added logic for handling v >= 999, we handle with 1 more precision , but I didn't for v< 999 otherwise it's kind of breaking experience for many users seeing too much detailed information.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs [hunk 2]
```diff
@@ -64,15 +64,17 @@ public static string NumberToString(long number, IFormatProvider culture)
 
             while (v >= 1000)
             {
-                v /= 1000;
+                v = Math.Round(v / 1000, 2);
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs]: On PMUI we don't see 3 digit precision, most 2 digits like below, otherwise previous logic having too detailed info causes problem.
image

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs (hunk 1 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs]: Isn't showing a 4th digit of precision when all other cases only show 3 also a kind of "breaking" experience?

How about `if (v > 999.0) { v = 999.0; }`?

Comment (erdembayar) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs]: @zivkan 
Ok. Taken your suggestion. Please check now.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/ConverterTests.cs [hunk 3]
```diff
@@ -21,12 +25,50 @@ public class ConverterTests
...
-        [InlineData(1234000000, "1.23G")]
-        public void DownloadCountToStringTest(int num, string expected)
+        [InlineData(123_400, "123K")]
+        [InlineData(899_560, "900K")]
+        [InlineData(998_999, "999K")]
+        [InlineData(999_000, "999K")]
+        [InlineData(999_001, "999K")]
+        [InlineData(999_560, "999K")]
+        [InlineData(1_234_000, "1.23M")]
+        [InlineData(1_299_560, "1.3M")]
```
Comment (zivkan) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/ConverterTests.cs:40]: From a scientific point of view, 1.3 and 1.30 mean different things. Should we fix this?

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/ConverterTests.cs [hunk 4]
```diff
@@ -21,12 +25,50 @@ public class ConverterTests
...
+        [InlineData(9_996_250, "10M")]
+        [InlineData(9_995_650, "10M")]
```
Comment (zivkan) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/ConverterTests.cs]: These tests seem so similar, I'm not sure why they're considered testing different scenarios/edge cases.

Below there's also a bunch of other test cases whose numbers are so similar I don't understand the intent of having multiple tests. The 100T tests in particular, the difference is 6.0e2 in a 9.99e13 number that gets rounded off to 3 significant digits. It's not testing the difference between rounding up vs down.

**Review by zivkan — APPROVED** [review]

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/ConverterTests.cs (hunk 3 above)
Comment (erdembayar) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/ConverterTests.cs:40]: This case not affected by change it's same as previous experience, for now I don't believe users really pay that amount of attention into it. If it arises in the future then we'll fix it that one.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/ConverterTests.cs (hunk 4 above)
Comment (erdembayar) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/ConverterTests.cs]: Removed duplicated ones.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs [hunk 5]
```diff
@@ -47,8 +47,8 @@ private static bool IsHttpUrl(Uri uri)
             return (uri.Scheme == Uri.UriSchemeHttp || uri.Scheme == Uri.UriSchemeHttps);
         }
 
-        private static readonly string[] _scalingFactor = new string[] {
-            String.Empty,
+        private static readonly string[] ScalingFactor = new string[] {
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs]: Shouldn't this stay as `_scalingFactor`?
https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md#c-coding-style

**Review by donnie-msft — APPROVED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs (hunk 5 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs]: This is IDE suggested fix, it looks rule is bit different for `static readonly` variable.

https://github.com/NuGet/NuGet.Client/blob/f4ae349ec18d1ac5d5e7d1b0ed753e1300e9e157/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs#L42

Comment (jeffkl) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs]: I think this whole algorithm is too complex and there's a bug here where billions will look like `1.23G` because `G` is specified for billions instead of `B`.  Does this scaling factor need to know about byte size and item count?  If so then it would need to differentiate them.

I think we should simplify this method and get rid of the loop which is confusing and made you add this silly looking if statement:
```
// If number is 999561 then v = 999.56 so with G3 which result in 1000k ~ 1E+03K.
if (v > 999)
{
    v = 999;
}
```

What do you think about this instead?
```C#
public static string NumberToString(long number, IFormatProvider culture)
{
    const double thousand = 1000D;
    const double million = 1000000D;
    const double billion = 1000000000D;
    const double trillion = 1000000000000D;

    if (number >= trillion)
    {
        return (number / trillion).ToString("0.##T", culture);
    }

    if (number >= billion)
    {
        return (number / billion).ToString("0.##B", culture);
    }

    if (number >= million)
    {
        return (number / million).ToString("0.##M", culture);
    }

    if (number >= thousand)
    {
        return (number / thousand).ToString("0.##K", culture);
    }

    return number.ToString("G0", culture);
}
```

Then you can get rid of the `ScalingFactor` array as well and this method makes waaaaay more sense.

We might need to methods:
`ByteCountToString()`: Returns the number as bytes so KB, MB, GB, TB (kilobytes, megabytes, gigabytes, terabytes)
`CountToString()`: Returns a number as a count so K, M, B, T (thousand, million, billion, trillion)

**Review by jeffkl — CHANGES_REQUESTED** [review]
Feel  free to ping me to discuss if you want, I really think we should fix this code as a whole

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs (hunk 5 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs]: > because G is specified for billions instead of B.

G is the SI (standard international) prefix. If we want to use B in English, then we need to make it a string resource to allow the translators to localize this into the correct value for each language.  So that part of it is by-design, I think.  See also https://github.com/NuGet/Home/issues/10864

Comment (jeffkl) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs]: Well yeah while we're here we might as well localize it!  Fix two bugs with one stone

**Review by aortiz-msft — APPROVED** [review]
Approved for insertion into dev branch.

**Comment by erdembayar** [comment]
@jeffkl 
Yes, I thought same approach as yours. But I didn't want to make breaking change this time, let's keep it as it's now. 
With `0.##K` format it introduces extra precision compare to previous experience, we better to consult PM before making any breaking change.
image

We previously had discussion on replacing `G` with `B` then decided [not to act](https://github.com/NuGet/Home/issues/10864#issuecomment-863543523 ) that time. For people on other side of world current one is [correct](https://github.com/NuGet/Home/issues/10864#issuecomment-863589046).
Also as far I can find only used with download count, it's not used with size conversion with bytes, so not concern for now.

**Comment by jeffkl** [comment]
I don't think changing the precision would be a breaking change but we could keep it as-is with `0.#K` right?  We can localize the G/B in a subsequent commit but I think we need to simplify this method, especially if the unit test coverage is great which it looks like is based on your added test case data.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs (hunk 5 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs]: @jeffkl @zivkan 
I took Jeff's code as base and added localization text, please check now.
Number - Previous experience - New experience
image

**Review by jeffkl — APPROVED** [review]
Looks good to me, thanks for localizing it too. Update the PR description to say it fixes https://github.com/NuGet/Home/issues/10864 ?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs [hunk 6]
```diff
@@ -47,33 +47,51 @@ private static bool IsHttpUrl(Uri uri)
...
+            {
+                double val = num / precision;
+                return val > 999 ? 999 : val;
+            }
+
+            string GetFormatted(double num, string precisionString)
+            {
+                return string.Format(
+                   culture,
+                   "{0:G3}{1}",
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs]: Wouldn't it be better to include the `{0:G3}` in the localized string?  Maybe some cultures prefer to have a space between the number and the suffix, while others don't. For example `123 M` vs `123M`

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Resources.resx [hunk 7]
```diff
@@ -1005,4 +1005,20 @@ Please see https://aka.ms/troubleshoot_nuget_cache for more help.</value>
...
+    <value>B</value>
+    <comment>1.000.000.000 in short format notation</comment>
+  </data>
+  <data name="Million" xml:space="preserve">
+    <value>M</value>
+    <comment>1.000.000 in short format notation</comment>
+  </data>
+  <data name="Thousand" xml:space="preserve">
+    <value>K</value>
+    <comment>1.000 in short format notation</comment>
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Resources.resx]: English uses `,` for thousands separators, so while all of these comments use `.` which is probably more common across the world, given our code and comments are in English, it looks a bit odd, and this one (thousands) can be mistaken for the number one (precise to 3 decimal places).

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs (hunk 6 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs]: @zivkan 
I moved into localized string, please review again.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Resources.resx [hunk 8]
```diff
@@ -1005,4 +1005,20 @@ Please see https://aka.ms/troubleshoot_nuget_cache for more help.</value>
     <value>This package version is deprecated. Use {0} instead.</value>
     <comment>{0} is a NuGet package id (string)</comment>
   </data>
+  <data name="Billion" xml:space="preserve">
+    <value>{0:G3}B</value>
+    <comment>Billions in short format notation</comment>
```
Comment (dominoFire) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Resources.resx]: Please specify which billions are you referring to:

```suggestion
    <comment>American billions in short format notation</comment>
```

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Resources.resx [hunk 9]
```diff
@@ -1005,4 +1005,20 @@ Please see https://aka.ms/troubleshoot_nuget_cache for more help.</value>
...
+    <value>{0:G3}M</value>
+    <comment>Millions in short format notation</comment>
+  </data>
+  <data name="Thousand" xml:space="preserve">
+    <value>{0:G3}K</value>
+    <comment>Thousands in short format notation</comment>
+  </data>
+  <data name="Trillion" xml:space="preserve">
+    <value>{0:G3}T</value>
+    <comment>Trillions in short format notation</comment>
```
Comment (dominoFire) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Resources.resx]: Please specify which trillions are you referring to:

```suggestion
    <comment>American trillions in short format notation</comment>
```

**Review by dominoFire — COMMENTED** [review]
Some comments on strings. I just want to make sure loc team will not get confused with two different meanings in the same string resource

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Resources.resx (hunk 9 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Resources.resx]: @dominoFire 
Does EU and US trillions use different notations?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Resources.resx (hunk 8 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Resources.resx]: Addressed.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Resources.resx (hunk 9 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Resources.resx]: https://en.wikipedia.org/wiki/Long_and_short_scale

Comment (erdembayar) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Resources.resx]: > https://en.wikipedia.org/wiki/Long_and_short_scale

Thanks you. Now I understand what `milliard` actually means, previously I always thought it's just another regional spelling for million.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/ConverterTests.cs [hunk 10]
```diff
@@ -21,14 +25,48 @@ public class ConverterTests
...
+        [InlineData(999_999_999_650, "999B")]
+        [InlineData(9_999_999_999_650, "10T")]
+        [InlineData(19_999_999_999_050, "20T")]
+        [InlineData(99_999_999_999_650, "100T")]
+
+        // there is localization
+        [InlineData(1939, "1.94K", "en-US")]
+        [InlineData(1939, "1,94K", "da-DK")]
+        [InlineData(9456, "9.46K", "en-US")]
+        [InlineData(9456, "9,46K", "da-DK")]
```
Comment (aortiz-msft) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/ConverterTests.cs:65]: >[InlineData(9456, "9,46K", "da-DK")] [](http://example.com/codeflow?start=8&length=36)

Could we also add tests for B vs G, since Andy pointed out that G will be used for locales following the metric system?

Comment (erdembayar) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/ConverterTests.cs:65]: @aortiz-msft
I thought about it but I believe it's available after localization team make translation for each different cultures. @dominoFire Please correct me if I'm wrong.

Comment (zivkan) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/ConverterTests.cs:65]: Also, since NuGet.Client's localization happens in a different repo, we can't have unit tests that work the same and still pass both locally (that doesn't run MicroBuild's localization task) and on CI (which does run MicroBuild's localization task).

I believe that if we migrated to the same localization system/format that the rest of the dotnet org uses, then local builds would include it.
